### ISSUE TITLE
RoBERTa/BART: Classification heads without fine-tuning

### ISSUE THREAD

**Issue by prihoda** [pytorch/fairseq#1663]
Hi, I'm a bit confused about classification heads in RoBERTa/BART. Is there an option to disable fine-tuning and train just using the frozen features? Or if only fine-tuning is possible at this moment, what is the point of having multiple classification heads?

Thanks!

**Comment by lematt1991** [comment]
> Is there an option to disable fine-tuning and train just using the frozen features?

I'm not sure I follow, what do you mean train with just the frozen features?  What parameters do you want to train?

**Comment by prihoda** [comment]
@lematt1991 "frozen" was probably a bad choice of words. I am looking for two things:
1) Is it possible to freeze the encoder weights and train just the classification head?
2) Is it possible to pre-compute the outputs of the transformer encoder for my (small) training set so that they don't have to be recomputed on each epoch when training just the classification head?

**Comment by lematt1991** [comment]
At the moment, I don't believe that either of these are supported, although, freezing the `sentence_encoder` layers should be easy.  

1. Add a `--freeze-encoder-weights` flag [here](https://github.com/pytorch/fairseq/blob/master/fairseq/tasks/sentence_prediction.py#L60)
2. Add the following [here](https://github.com/pytorch/fairseq/blob/master/fairseq/tasks/sentence_prediction.py#L212):

```Python
if args.freeze_encoder_weights:
    for param in model.model.decoder.sentence_encoder.parameters():
        param.requires_grad = False
```

**Comment by prihoda** [comment]
Great, thanks! 

And while we're on it, I will also be looking at _per-token_ classification using RoBERTa, AFAIK this is also not supported at the moment, right? The final LM layer is per-token, but the output is fixed to the same dictionary as the input, so it cannot be reused for classification. So I guess it would make sense to:

- Add a `--classify-per-token` flag to the sentence prediction task, which would tell the classification head to process all features, not just the `<s>` token. Do you think I should add LayerNorm as in the LM?
- Adapt the sentence prediction criterion to support per-token classification. Should this look for the `classify_per_token` flag or deduce it from the shape of the model outputs? Or should I adapt e.g. the `cross_entropy` loss instead by adding a `--classification-head-name` arg?

**Comment by prihoda** [comment]
Creating a separate issue for a related feature: https://github.com/pytorch/fairseq/issues/1708

**Comment by prihoda** [comment]
Adding feature request details

## 🚀 Feature Request
E

### Motivation



### Pitch



### Alternatives



### Additional context
